test/distributed/checkpoint/e2e/test_fine_tuning.py

```diff
@@ -0,0 +1,192 @@
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+DIM = 500
+
+
+class PreTrainedModel(nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.layer1 = nn.Linear(DIM, DIM)
+        self.layer2 = nn.Linear(DIM, DIM)
+        self.layer3 = nn.Linear(DIM, DIM)
+        self.relu = nn.ReLU()
+
+    def forward(self, batch):
+        x = self.relu(self.layer1(batch))
+        x = self.relu(self.layer2(x))
+        x = self.relu(self.layer3(x))
+        return x
+
+
+class FineTuningModel(nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.pretrain = PreTrainedModel()
+        for p in self.pretrain.parameters():
+            p.requires_grad = False
+
+        self.layer1 = nn.Linear(DIM, DIM)
+        self.layer2 = nn.Linear(DIM, DIM)
+        self.layer3 = nn.Linear(DIM, DIM)
+        self.relu = nn.ReLU()
+
+    def forward(self, batch):
+        x = self.relu(self.pretrain(batch))
+        x = self.relu(self.layer1(x))
+        x = self.relu(self.layer2(x))
+        x = self.relu(self.layer3(x))
+        return x
+
+
+class TestFineTuning(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(4, torch.cuda.device_count())
+
+    def pretrain(self, pretrain_dir: str) -> None:
+        model = PreTrainedModel().cuda()
+        model = FSDP(model)
+        optim = torch.optim.Adam(model.parameters(), lr=1e-3)
+
+        # Trainining
+        for i in range(3):
+            batch = torch.rand(32, DIM, device="cuda")
+            loss = model(batch).sum()
+            loss.backward()
+            optim.step()
+            optim.zero_grad()
+
+        # Save state_dict
+        model_state_dict, optim_state_dict = get_state_dict(model, optimizers=optim)
+        saved_state_dict = {"model": model_state_dict, "optim": optim_state_dict}
+        dist_cp.save_state_dict(
+            state_dict=saved_state_dict,
+            storage_writer=dist_cp.FileSystemWriter(pretrain_dir),
+        )
+
+    def finetune(self, pretrain_dir: str, finetune_dir: str) -> None:
+        model = FineTuningModel().cuda()
+        # TODO: make the parallelism more complicated, e.g., using 2D + DDP.
+        model = FSDP(model, use_orig_params=True)
+        optim = torch.optim.Adam(model.parameters(), lr=1e-3)
+
+        # Simulate that the fine tuning restart after 3 iterations
+        for i in range(2):
+            # Load pretrain submodules checkpoint
+            pretrain_state_dict, _ = get_state_dict(
+                model,
+                submodules={model.pretrain},
+                options=StateDictOptions(keep_submodule_prefixes=False)
```

Comment:
what does `keep_submodule_prefixes` mean? How would a user decide / know whether to use this flag or not?

Reply:
This is a good question. It mean whether the keys of  the returned state_dict should keep the submodule prefixes. For this example, if we set the flag to True (the default one), the returned state_dict will be `{"pretrain.layer1.weight" ...}`. And if we set the flag to False, the returned state_dict will be `{"layer1.weight", ...}`.

How would a user knows? Unfortunately, this depends on what is saved in the checkpoint. If the checkpoint already has the prefix, the user should set it to True otherwise False.